productinfo/ec2/productinfo_ec2.go

```diff
@@ -125,27 +126,42 @@ func (e *Ec2Infoer) GetProducts(regionId string, attrKey string, attrValue produ
 	if err != nil {
 		return nil, err
 	}
-	for _, price := range products.PriceList {
-		var onDemandPrice float64
-		// TODO: this is unsafe, check for nil values if needed
-		instanceType := price["product"].(map[string]interface{})["attributes"].(map[string]interface{})["instanceType"].(string)
-		cpusStr := price["product"].(map[string]interface{})["attributes"].(map[string]interface{})[Cpu].(string)
-		memStr := price["product"].(map[string]interface{})["attributes"].(map[string]interface{})[Memory].(string)
-		var gpus float64
-		if price["product"].(map[string]interface{})["attributes"].(map[string]interface{})["gpu"] != nil {
-			gpuStr := price["product"].(map[string]interface{})["attributes"].(map[string]interface{})["gpu"].(string)
-			gpus, _ = strconv.ParseFloat(gpuStr, 32)
+	for i, price := range products.PriceList {
+		pd, err := newPriceData(price)
+		if err != nil {
+			log.Warn("could not extract pricing info for the item with index: [ %d ]", i)
+			continue
 		}
-		onDemandTerm := price["terms"].(map[string]interface{})["OnDemand"].(map[string]interface{})
-		for _, term := range onDemandTerm {
-			priceDimensions := term.(map[string]interface{})["priceDimensions"].(map[string]interface{})
-			for _, dimension := range priceDimensions {
-				odPriceStr := dimension.(map[string]interface{})["pricePerUnit"].(map[string]interface{})["USD"].(string)
-				onDemandPrice, _ = strconv.ParseFloat(odPriceStr, 32)
-			}
+
+		instanceType, err := pd.GetInstanceType()
+		if err != nil {
+			log.Warn("could not get instance type")
+			return nil, err
```

Comment:
don't exit the processing in case of these errors, just use `continue` instead - there might be valid entries in case one is broken